torchtitan/optimizer.py

```diff
@@ -40,6 +41,45 @@ def _build_optimizer(model):
 
         return optimizer
 
+    def _build_optimizer_in_backward(model):
+        name = job_config.optimizer.name
+        lr = job_config.optimizer.lr
+        fused = job_config.optimizer.fused
+
+        # Common parameters for both optimizers
+        optimizer_kwargs = {
+            "lr": lr,
+            "betas": (0.9, 0.95),
+            "weight_decay": 0.1,
+            "fused": fused,
+            "foreach": not fused,
+        }
+        if name == "Adam":
+            # TODO: make the optimizer options configurable by toml/cmd args
+            optim_dict = {
+                param: torch.optim.Adam([param], **optimizer_kwargs)
+                for param in model.parameters()
+            }
+        elif name == "AdamW":
+            optim_dict = {
+                param: torch.optim.AdamW([param], **optimizer_kwargs)
+                for param in model.parameters()
+            }
+        else:
+            raise NotImplementedError(f"Optimizer {name} not added.")
+
+        def optim_hook(param) -> None:
+            optim_dict[param].step()
+            optim_dict[param].zero_grad()
+
+        for param in model.parameters():
+            if param.requires_grad:
+                param.register_post_accumulate_grad_hook(optim_hook)
```

Comment:
We probably need to document that users could not call `register_post_accumulate_grad_hook` after the optimizer is built. Otherwise, the gradients are clear and users' `register_post_accumulate_grad_hook` will be useless.

Reply:
Added caution in config_manager